### DISCUSSION TITLE
Argon development discussion

### DISCUSSION THREAD

**Discussion by DervexDev** [argon-rbx/argon#24] [General]
### Found a bug? Have some suggestions?

If so, this is the perfect place to comment!

**Comment by EthanMichalicek** [comment]
My coworker @kolson25 and I have been converting our large project from Argon 1 > Argon 2. We've compiled a list of roadblocks we've run into along the way. For now, we're working around all of these in one way or another, but having these as supported features would be amazing. 🙏 

- A String Value that was a child of a Model in ReplicatedStorage consistently doesn't get replicated in VS Code/git. Other instances of String Values did get replicated as `*.txt` files. The full path to the String Value was: `ReplicatedStorage/Town/Arena/TrackTemplate/ID` (`TrackTemplate` is a Model; `ID` is the String Value). ~~As I was typing this, we ran into another issue with a BoolValue in ReplicatedStorage that is also struggling to sync.~~ As I was typing _that_, it seems that it was not limited to ReplicatedStorage - we just had the same naming issue appear in the workspace as well.
- Dealing with Packages in Argon is a bit weird at the moment. Two issues worth noting:
  1. `PackageLink` is not a spawnable instance. This itself is not a problem, but does sometimes result in a warning.
  2. Whenever Argon syncs a Package from VS Code to Studio (right after Argon is connected), Studio sees all of the file syncing in the Package's hierarchy as changes, even if nothing changed. This results in a new popup window window spawning per change asking if we want to accept the change or revert it to the original. In some cases, this is tens or hundreds of windows spawned on top of each other.
  3. Ideally, we'd like an option to ignore syncing Packages entirely. If we use the `ignoreGlob` feature, we can use wildcards to ignore the Packages (by using `ignoreGlob: ["**/PackageLink/.."]`), but then it deletes the Package from our workspace every time we connect Argon. Because of this, we're instead loading all of our packages on server load instead of dropping them in the workspace.
- Sibling objects with the same name in a given hierarchy will only have one instance synced to VS Code/git. For now, we're working around this by giving all siblings unique names. An option to generate IDs or use some other unique identifier might work to fix this issue.
- Attributes and Tags do not seem to get synced at all. It's possible we're doing something wrong here, but in our tests so far they have not appeared in VS Code.

For reference, we're using the following settings:
![image](https://github.com/argon-rbx/argon/assets/51338441/601aad51-a464-48b5-bd2c-eb721771c443)

**Reply by DervexDev** [reply]
Thanks for the detailed list of problems! I will start working on fixing all of these issues when I have some free time. I have some initial notes though.

> A String Value that was a child of a Model in ReplicatedStorage consistently doesn't get replicated in VS Code/git. Other instances of String Values did get replicated as *.txt files.

Make sure it doesn't get created in your project file.

> Whenever Argon syncs a Package from VS Code to Studio (right after Argon is connected), Studio sees all of the file syncing in the Package's hierarchy as changes, even if nothing changed. This results in a new popup window window spawning per change asking if we want to accept the change or revert it to the original. In some cases, this is tens or hundreds of windows spawned on top of each other.

I honestly didn't know that something like `PackageLink` ever existed thus Argon can't manage them correctly as they are "special" instances which need custom handling.

> Ideally, we'd like an option to ignore syncing Packages entirely. If we use the ignoreGlob feature, we can use wildcards to ignore the Packages (by using ignoreGlob: ["**/PackageLink/.."]), but then it deletes the Package from our workspace every time we connect Argon. Because of this, we're instead loading all of our packages on server load instead of dropping them in the workspace.

For now, you can enable the `Skip Initial Sync` setting. It makes Argon 2 behave almost the same as Argon 1 as it no longer checks whether the client and server are in sync.

> Sibling objects with the same name in a given hierarchy will only have one instance synced to VS Code/git. For now, we're working around this by giving all siblings unique names. An option to generate IDs or use some other unique identifier might work to fix this issue.

Argon 1 assignes UUID to every duplicate instance but I don't think this is a good solution. Instead, I would like to add the ability to mark selected instances to be converted to [JSON Models](https://argon.wiki/api/file-types#json-model) which support duplicates.

> Attributes and Tags do not seem to get synced at all. It's possible we're doing something wrong here, but in our tests so far they have not appeared in VS Code.

Just tested and both Tags and Attributes sync properly. Could you provide more details on this problem?

Also, your `*.project.json` file would be very useful to solve some of these issues.

**Comment by kolson25** [comment]
[town.project.json](https://github.com/user-attachments/files/15566757/town.project.json)
This is our project.json we were testing with. Can hopefully get you some follow up details soon.  I do know the tags and attributes were being added in Workspace to a Part when were weren't seeing them syncing.

**Reply by EthanMichalicek** [reply]
As some added context for this project file:
- Everything in our Roblox place file is placed either in one of the `Global` or `Town` folders.
- Because we turned on property syncing, the missing String Values did not end up in the project file; the file linked above remained static once we turned on that Argon feature.
- `StarterPlayerScripts` and `StarterCharacterScripts` break our format _slightly_ because they have to remain as direct children of `StarterPlayer` to function.
- `Town` and `Global` in `StarterGui`, despite being paths, have their `.data.json` class names set as `ScreenGui` because placing ScreenGuis subfolders breaks certain features in the Roblox GUI system.

**Reply by DervexDev** [reply]
Is the problem with attributes related to #27?

**Reply by EthanMichalicek** [reply]
Doesn't seem to be - we only ever tried creating tags or attributes in Studio itself (with Argon running). The project json was not modified, but nothing was added to the folders in VS Code either (in most attempts; I think we saw a .txt file appear once or twice).

**Comment by ghost** [comment]
Can someone elaborate what's the difference between an argon sourcemap and a rojo sourcemap is? Because somehow argon isn't compatible with darklua, but rojo is!

**Reply by DervexDev** [reply]
As discussed on the Roblox OSS Community after a small fix the only difference is that Argon uses absolute paths, unlike Rojo which uses relative. There will be a new option to change it in the global config.

**Reply by DervexDev** [reply]
Small change - `relative_paths` setting has been removed because of the problems it created. Now sourcemap uses relative paths by default and fallbacks to the absolute ones if needed.

**Comment by DervexDev** [comment]
@EthanMichalicek and @kolson25 I just started working on fixing the issues you listed and here is the current status:
- Problem with `StringValue` under the path you provided - can't reproduce, need a minimal repro ❌
- General problem with `BoolValues` - can't reproduce, need a minimal repro ❌
- `PackageLinks` - are now supported, and available in [2.0.9](https://github.com/argon-rbx/argon/releases/tag/2.0.9) ✅
- Instances with the same name - new way to mark instances to be transformed into `JSON Models`, scheduled for Argon `2.1` ⏺️
- Problem with `Tags` and `Attributes` -  only affects `Value` like instances, fixed in [2.0.8](https://github.com/argon-rbx/argon/releases/tag/2.0.8) ✅

**Reply by EthanMichalicek** [reply]
Good to hear! Unfortunately we're really busy with the project at the moment and don't have time we can dedicate to reproducing the issues in a siloed example project, but if we happen across them again in a more easily shareable manner, we'll send it right your way.

**Comment by AridAjd** [comment]
Not sure if I'm using it wrong but I can't set `Lighting.Technology` through my `project.json` file? I tried in both implicit and explicit formats:

![image](https://github.com/argon-rbx/argon/assets/118916772/8a489860-ba6b-47de-a021-56e9c243c4a2)
---
![image](https://github.com/argon-rbx/argon/assets/118916772/f68bcb06-fb80-4edb-9c70-9a8551b4fe69)

I also tried changing other props and they change and they are shown in the output window, however the `Technology` property doesn't even appear here.
![image](https://github.com/argon-rbx/argon/assets/118916772/252afc84-ffae-4c22-8bf5-30a08d99365e)

**EDIT**: Just realized that you have to put the value of the enum, which I also tried but it still didn't work.

**Reply by DervexDev** [reply]
Continue in #59

**Reply by AridAjd** [reply]
Added a repro to the issue 👍

**Comment by DervexDev** [comment]
### In Argon `2.0.11` you can finally sync MeshParts!
See the demo [here](https://youtu.be/wRgGfc0W-k0)
